### PR TITLE
Floating-point support in `fud`

### PR DESCRIPTION
**Pull request by rachitnigam**
Extract floating-point support for `fud` from https://github.com/calyxir/calyx/pull/1928

### CODE REVIEW

File: tests/correctness/float.futil.data
Comment (ekiwi) [inline tests/correctness/float.futil.data:14]: should this be  `15.1239`?

File: tests/correctness/float.futil.data
Comment (ekiwi) [inline tests/correctness/float.futil.data:15]: should this be `0.56`?

Comment (ekiwi) [reply tests/correctness/float.futil.data:15]: I got it, this is the input, not the expect file. My bad.

File: tests/correctness/float.futil.data
Comment (rachitnigam) [reply tests/correctness/float.futil.data:14]: This is the input file. The output file is the `.expect` file.

Comment (ekiwi) [reply tests/correctness/float.futil.data:14]: I got it, this is the input, not the expect file. My bad.

File: tests/correctness/float.futil.data
Comment (rachitnigam) [reply tests/correctness/float.futil.data:15]: Same as above

**Review by ekiwi — CHANGES_REQUESTED** [review]
Could you please also add the floating-point conversion to whatever data converter `cider` uses? We definitely do not want `cider` to diverge from the calyx compilation flow any more than it already has.

**Comment by rachitnigam** [comment]
Unfortunately it seems that cider uses a different data converter altogether. I'm going to open an issue and tackle that in a follow-up PR.

**Comment by rachitnigam** [comment]
Also, this doesn't support negative numbers so we need to fix that before merging this PR in.

**Comment by ekiwi** [comment]
> Unfortunately it seems that cider uses a different data converter altogether. I'm going to open an issue and tackle that in a follow-up PR.

Will you be responsible for adding Cider support in a follow-up PR? It can't just all be Griffin's job to update Cider. If Cider is supposed to be an integral part of the Calyx ecosystem, then new features need to be added to both, the compiler and Cider at the same time.

**Comment by rachitnigam** [comment]
> Will you be responsible for adding Cider support in a follow-up PR?

Perhaps! But I can't commit to doing it in the next couple of weeks. The background on this PR is that we need floating-point support for @andrewb1999's work which has a hard deadline. In general, when we have such deadline pushes, we tend to open PRs that fix the narrow path and open issues to address missing features in the future.

> If Cider is supposed to be an integral part of the Calyx ecosystem, then new features need to be added to both, the compiler and Cider at the same time.

I agree with you in theory but the amount of separation in the two ecosystems makes this tricky. For example, the cost of adding a new primitive to Cider is quite high (see #2030 for a recent example). Similarly, the data converter being a separate entity means that I would have to make sure that the changes really make sense for Cider. For the compiler, we can just pull in hardfloat but for cider, the "right way" to support floating-point stuff does not seem that obvious to me yet (do we add new primitives, BTOR support, co-simulation?).

> It can't just all be Griffin's job to update Cider.

Certainly! One person maintaining the whole ecosystem is a pretty big ask. Unfortunately that is currently my position but it extends to the entire infrastructure; for example, there is currently no designated maintainer for `fud` so I am the one shepherding these changes. The goal of this PR is to unblock Andrew, and therefore I am okay with it being a narrowly scoped. In a world where workload, initiative, and management of the project was more equitably distributed, I might be able to sit down and do this the right way instead of putting out individual fires as they come up. As things stand, this is not the case and I have to get the things I can done in the time I have. 

Finally, if convincing is needed on me helping maintain various parts of the ecosystem, the following command will provide a simple counterexample in any directory in this repo:
```bash
% git shortlog -sn -- interp | head -n 3  # number of commits grouped by author
   101  Griffin Berlstein
    80  Rachit Nigam
    12  Karen Z
```

**Comment by jiahanxie353** [comment]
> Also, this doesn't support negative numbers so we need to fix that before merging this PR in.

You mean https://github.com/calyxir/calyx/pull/2316 did not take negative numbers into account or?

**Review by ekiwi — APPROVED** [review]
OK. Let's get this in to make sure Andrew hits his deadline.

**Comment by rachitnigam** [comment]
> You mean https://github.com/calyxir/calyx/pull/2316 did not take negative numbers into account or?

Ah, you're right. This is a problem with that older PR...

**Pull request merged**
